rfd/0167-console-service.md

```diff
@@ -0,0 +1,253 @@
+---
+authors: Gabriel Corado (gabriel.oliveira@goteleport.com)
+state: draft
+---
+
+# RFD 0167 - Debug service
+
+## Required Approvals
+
+* Engineering: @r0mant && (@greedy52 || @rosstimothy)
+* Security: @reedloden || @jentfoo
+
+## What
+
+This RFD proposes a mechanism for Teleport administrators to switch log-level
+and enable profile consumption without restarting the instance.
+
+## Why
+
+During incidents in production environments, debug logs and diagnostic profiles
+(such as CPU and Memory allocation) are often unavailable for timely
+troubleshooting. To enable those, administrators need to restart the instance,
+potentially disrupting the application state and obscuring the issue they're
+trying to solve.
+
+In addition, the addition will make collecting profiles with a single command
+even on distroless environments where `curl` is not present.
+
+## Details
+
+A separate service will be introduced specifically for troubleshooting
+purposes. This will ensure the current diagnosis service is kept as is, not
+breaking current integrations or introducing behavior changes. The new service
+will also always be enabled by default (contrasting with the diagnosis service,
+which is optional and disabled by default), so in scenarios where users need to
+use it, they don't need to restart Teleport.
+
+In addition, the new service will listen into a Unix socket instead of TCP. This
+will make discoverability easier and discourage external usage, as the API is
+designed for internal usage.
+
+The `pprof` endpoints on diagnostics will still be available for existent
+integrations, such as the usage of
+`go tool pprof http://diag-addr/debug/pprof/profile`.
+
+### New service
+
+Teleport will start listening using a Unix socket located at
+`<data_dir>/debug.sock`.
```

Comment:
It would be _really_ nice to also have a socket at `<pid>.sock` (or maybe `<pid>.<TeleportProcess unique identifier>.sock`? not sure if we intend to serve healthz and readyz from the socket, those would be specific to the instance of `TeleportProcess`) for slightly more advanced debugging in situations with multiple Teleport agent processes running.

Reply:
(that being said, it's also possible to do this manually by just linking the socket before doing any sort of restart, so for development purposes this is not strictly necessary, and users might get confused by it)